src/vmm/src/vstate/kvm.rs

```diff
@@ -0,0 +1,219 @@
+    GetSupportedCpuId(kvm_ioctls::Error),
+    /// The number of configured slots is bigger than the maximum reported by KVM
+    NotEnoughMemorySlots,
+}
+
+/// Struct with kvm fd and kvm associated paramenters.
+#[derive(Debug)]
+pub struct Kvm {
+    /// KVM fd.
+    pub fd: KvmFd,
+    /// Maximum number of memory slots allowed by KVM.
+    pub max_memslots: usize,
+    /// Additional capabilities that were specified in cpu template.
+    pub kvm_cap_modifiers: Vec<KvmCapability>,
+
+    // X86 specific fields.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    /// Supported CpuIds.
+    pub supported_cpuid: CpuId,
+    #[cfg(target_arch = "x86_64")]
+    /// Msrs needed to be save on snapshot creation.
+    pub msrs_to_save: MsrList,
+}
+
+impl Kvm {
+    /// Create `Kvm` struct.
+    pub fn new(kvm_cap_modifiers: Vec<KvmCapability>) -> Result<Self, KvmError> {
+        let kvm_fd = KvmFd::new().map_err(KvmError::Kvm)?;
+
+        // Check that KVM has the correct version.
+        // Safe to cast because this is a constant.
+        #[allow(clippy::cast_possible_wrap)]
+        if kvm_fd.get_api_version() != KVM_API_VERSION as i32 {
+            return Err(KvmError::ApiVersion(kvm_fd.get_api_version()));
+        }
+
+        let total_caps = Self::combine_capabilities(&kvm_cap_modifiers);
+        // Check that all desired capabilities are supported.
+        Self::check_capabilities(&kvm_fd, &total_caps).map_err(KvmError::Capabilities)?;
+
+        let max_memslots = kvm_fd.get_nr_memslots();
+
+        #[cfg(target_arch = "aarch64")]
+        {
+            Ok(Self {
+                fd: kvm_fd,
+                max_memslots,
+                kvm_cap_modifiers,
+            })
+        }
+
+        #[cfg(target_arch = "x86_64")]
+        {
+            let supported_cpuid = kvm_fd
+                .get_supported_cpuid(KVM_MAX_CPUID_ENTRIES)
+                .map_err(KvmError::GetSupportedCpuId)?;
+            let msrs_to_save = crate::arch::x86_64::msr::get_msrs_to_save(&kvm_fd)
+                .map_err(KvmError::GetMsrsToSave)?;
+
+            Ok(Kvm {
+                fd: kvm_fd,
+                max_memslots,
+                kvm_cap_modifiers,
+                supported_cpuid,
+                msrs_to_save,
+            })
+        }
+    }
+
+    /// Check guest memory does not have more regions than kvm allows.
+    pub fn check_memory(&self, guest_mem: &GuestMemoryMmap) -> Result<(), KvmError> {
+        if guest_mem.num_regions() > self.max_memslots {
+            Err(KvmError::NotEnoughMemorySlots)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn combine_capabilities(kvm_cap_modifiers: &[KvmCapability]) -> Vec<u32> {
+        let mut total_caps = Self::DEFAULT_CAPABILITIES.to_vec();
+        for modifier in kvm_cap_modifiers.iter() {
+            match modifier {
+                KvmCapability::Add(cap) => {
+                    if !total_caps.iter().any(|c| c == cap) {
```

Comment:
can't this be done with `.contains` instead? :thinking:

Reply:
fixed